src/libostree/ostree-sysroot-deploy.c

```diff
@@ -3478,6 +3478,65 @@ _ostree_sysroot_finalize_staged (OstreeSysroot *self,
   return TRUE;
 }
 
+/* Invoked at shutdown time by ostree-finalize-staged.service */
+gboolean
+_ostree_sysroot_finalize_staged (OstreeSysroot *self,
+                                 GCancellable  *cancellable,
+                                 GError       **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  if (!_ostree_sysroot_ensure_boot_fd (self, error))
+    return FALSE;
+  if (!_ostree_sysroot_finalize_staged_inner (self, cancellable, &local_error))
+    {
+      g_autoptr(GError) inner_error = NULL;
+      // Finalizing the deployment may have closed our boot_fd
+      if (!_ostree_sysroot_ensure_writable (self, &inner_error))
+        {
+          // For now...I think the best we can do here is just log.
+          g_printerr ("Failed to ensure sysroot writable: %s\n", inner_error->message);
+        }
+      else
+        {
+          g_assert_cmpint (self->boot_fd, !=, -1);
+          if (!glnx_file_replace_contents_at (self->boot_fd, _OSTREE_FINALIZE_STAGED_FAILURE_PATH, 
+                                               (guint8*)local_error->message, -1,
+                                               0, cancellable, &inner_error))
+            {
+              // We somehow failed to write the failure message...that's not great.  Maybe ENOSPC on /boot.
+              g_printerr ("Failed to write %s: %s\n", _OSTREE_FINALIZE_STAGED_FAILURE_PATH, inner_error->message);
+            }
+        }
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return FALSE;
+    }
+  return TRUE;
+}
+
+/* Invoked at bootup time by ostree-boot-complete.service */
+gboolean
+_ostree_sysroot_boot_complete (OstreeSysroot *self,
+                               GCancellable  *cancellable,
+                               GError       **error)
+{
+  if (!_ostree_sysroot_ensure_boot_fd (self, error))
+    return FALSE;
+
+  glnx_autofd int failure_fd = -1;
+  if (!ot_openat_ignore_enoent (self->boot_fd, _OSTREE_FINALIZE_STAGED_FAILURE_PATH, &failure_fd, error))
+    return FALSE;
+  // If we didn't find a failure log, then there's nothing to do right now.
+  // (Actually this unit shouldn't even be invoked, but we may do more in the future)
+  if (failure_fd == -1)
+    return TRUE;
+  g_autofree char *failure_data = glnx_fd_readall_utf8 (failure_fd, NULL, cancellable, NULL);
```

Comment:
Shouldn't this be

```suggestion
  g_autofree char *failure_data = glnx_fd_readall_utf8 (failure_fd, NULL, cancellable, error);
```

?

Reply:
Good catch!